File: src/js/3_StructuralPatterns/6_Flyweight/text_formatting.js
class FormattedText
{
  constructor(plainText)
  {
    this.plainText = plainText;
    this.caps = new Array(plainText.length).map(
      function() { return false; }
    );
  }

  capitalize(start, end)
  {
    for (let i = start; i <= end; ++i)
      this.caps[i] = true;
  }

  toString()
  {
    let buffer = [];
    for (let i in this.plainText)
    {
      let c = this.plainText[i];
      buffer.push(this.caps[i] ? c.toUpperCase() : c);
    }
    return buffer.join('');
  }
}

// this would work better as a nested class
class TextRange
{
  constructor(start, end)
  {
    this.start = start;
    this.end = end;
    this.capitalize = false;
    // other formatting options here
  }

  covers(position)
  {
    return position >= this.start &&
      position <= this.end;
  }
}

class BetterFormattedText
{
  constructor(plainText)
  {
    this.plainText = plainText;
    this.formatting = [];
  }

  getRange(start, end)
  {
    let range = new TextRange(start, end);
    this.formatting.push(range);
    return range;
  }

  toString()
  {
    let buffer = [];
    for (let i in this.plainText)
    {
      let c = this.plainText[i];
      for (let range of this.formatting) {
        if (range.covers(i) && range.capitalize)
          c = c.toUpperCase();
      }
      buffer.push(c);
    }
    return buffer.join('');
  }
}

const text = 'This is a brave new world';
let ft = new FormattedText(text);
ft.capitalize(10, 15);
console.log(ft.toString());

let bft = new BetterFormattedText(text);
bft.getRange(16, 19).capitalize = true;
console.log(bft.toString());

File: src/swift/3_StructuralPatterns/6_Flyweight/TextFormatting.swift
import Foundation

extension String {
    func substring(_ location: Int, _ length: Int) -> String? {
        guard characters.count >= location + length else { return nil }
        let start = index(startIndex, offsetBy: location)
        let end = index(startIndex, offsetBy: location + length)
        return substring(with: start..<end)
    }
}

class FormattedText: CustomStringConvertible
{
  private var text: String
  private var capitalize: [Bool]

  init(_ text: String)
  {
    self.text = text
    capitalize = [Bool](repeating: false, count: text.utf8.count)
  }
  
  func capitalize(_ start: Int, _ end: Int)
  {
    for i in start...end
    {
      capitalize[i] = true
    }
  }

  var description: String
  {
    var s = ""
    for i in 0..<text.utf8.count
    {
      let c = text.substring(i,1)!
      s += capitalize[i] ? c.capitalized : c
    }
    return s
  }
}

class BetterFormattedText: CustomStringConvertible
{
  private var text: String
  private var formatting = [TextRange]()

  init(_ text: String)
  {
    self.text = text
  }

  func getRange(_ start: Int, _ end: Int) -> TextRange
  {
    let range = TextRange(start, end)
    formatting.append(range)
    return range
  }

  var description: String
  {
    var s = ""
    for i in 0..<text.utf8.count
    {
      var c = text.substring(i, 1)!
      for range in formatting
      {
        if range.covers(i) && range.capitalize
        {
          c = c.capitalized
        }
      }
      s += c
    }
    return s
  }

  class TextRange
  {
    var start, end: Int
    var capitalize: Bool = false // bold, italic, etc

    init(_ start: Int, _ end: Int)
    {
      self.start = start
      self.end = end
    }

    func covers(_ position: Int) -> Bool
    {
      return position >= start && position <= end
    }
  }
}

func main()
{
  let ft = FormattedText("This is a brave new world")
  ft.capitalize(10,15)
  print(ft)

  let bft = BetterFormattedText("This is a brave new world")
  bft.getRange(10,15).capitalize = true
  print(bft)
}

main()